src/python/pants/backend/java/compile/javac.py

```diff
@@ -260,21 +265,41 @@ async def compile_java_source(
             use_nailgun=jdk_setup.digest,
             append_only_caches=jdk_setup.append_only_caches,
             output_directories=(dest_dir,),
-            description=f"Compile {request.component.members} with javac",
+            description=f"Compile {request.component} with javac",
             level=LogLevel.DEBUG,
         ),
     )
-    output: CompiledClassfiles | None = None
-    if process_result.exit_code == 0:
-        stripped_classfiles_digest = await Get(
-            Digest, RemovePrefix(process_result.output_digest, dest_dir)
+    if compile_result.exit_code != 0:
+        return FallibleCompiledClassfiles.from_fallible_process_result(
+            str(request.component),
+            compile_result,
+            None,
         )
-        output = CompiledClassfiles(stripped_classfiles_digest)
+
+    # Jar.
+    # NB: We jar up the outputs in a separate process because the nailgun runner cannot support
+    # invoking via a `bash` wrapper (since the trailing portion of the command is executed by
+    # the nailgun server). We might be able to resolve this in the future via a Javac wrapper shim.
```

Comment:
You may be able to use the `jar` utility here instead of directly calling `zip` -- it should be in there with the JDK.  _maybe_ nailgun will help there?

My main reason for not calling `jar` in the other PR was that I couldn't figure out what the classpath to the `jar` utility was, maybe you'll have more luck?

Reply:
Yea... the same issue remains though: invoking two consecutive Java processes in the same sandbox won't work as it stands (because back in the day I thought that CLI parsing would be the way to go, rather than having an explicit type for this... not sure yet whether that was a mistake, but we'll see).

It's possible that it could be made to work in the future by changing up how nailgun is used to maybe have the nailgun client be an explicit process inside the sandbox, but there would be a fair amount to figure out there. We'll see.